code/game/objects/structures/crematorium.dm

```diff
@@ -0,0 +1,254 @@
+	else if (open)
+		icon_state = "crematorium_open"
+	else if (contents.len > 1)
+		icon_state = "crematorium_filled"
+	else
+		icon_state = "crematorium_closed"
+
+/obj/structure/crematorium/explosion_act(severity)
+	..()
+	if(!QDELETED(src) && (severity == 1 || (severity == 2 && prob(50)) || (severity == 3 && prob(5))))
+		physically_destroyed()
+
+/obj/structure/crematorium/proc/open()
+	if(cremating || locked || open)
+		return
+
+	if(!connected_tray)
+		return
+
+	playsound(src, 'sound/items/Deconstruct.ogg', 50, 1)
+	var/turf/T = get_step(src, dir)
+	connected_tray.forceMove(T)
+	connected_tray.set_dir(dir)
+	for(var/atom/movable/A as mob|obj in src)
+		A.forceMove(get_turf(connected_tray))
+
+	open = TRUE
+	update_icon()
+
+/obj/structure/crematorium/proc/close()
+	if(!open)
+		return
+	
+	if(!connected_tray)
+		return
+
+	for(var/atom/movable/A as mob|obj in connected_tray.loc)
+		if(!A.anchored && !istype(A, connected_tray))
+			A.forceMove(src)
+
+	playsound(loc, 'sound/items/Deconstruct.ogg', 50, 1)
+	connected_tray.forceMove(src)
+
+	open = FALSE
+	update_icon()
+
+/obj/structure/crematorium/attack_hand(mob/user)
+	if(locked)
+		to_chat(usr, SPAN_WARNING("It's currently locked."))
+		return
+	
+	if(open)
+		close()
+	else
+		open()
+
+	return ..()
+
+/obj/structure/crematorium/attack_robot(mob/user)
+	if(Adjacent(user))
+		return attack_hand(user)
+	else 
+		return ..()
+
+/obj/structure/crematorium/attackby(P, mob/user)
+	add_fingerprint(user)
+	if(istype(P, /obj/item/pen))
+		var/new_label = sanitizeSafe(input(user, "What would you like the label to be?", capitalize(name), null) as text|null, MAX_NAME_LEN)
+
+		if((!Adjacent(user) || loc == user))
+			return
+
+		if(new_label)
+			SetName("[initial(name)] - \"[new_label]\"")
+		else
+			SetName(initial(name))
+
+	else 
+		return ..()
+
+/obj/structure/crematorium/relaymove(mob/user)
+	if (user.stat || locked)
+		return
+
+	open()
+
+/obj/structure/crematorium/proc/cremate(atom/A, mob/user)
+	if(cremating || open)
+		return //don't let you cremate something twice or w/e
+
+	if(contents.len <= 1)
+		audible_message(SPAN_WARNING("You hear a hollow crackle."), 1)
+		return
+
+	else
+		if(!isemptylist(search_contents_for(/obj/item/disk/nuclear)))
+			to_chat(loc, "The button's status indicator flashes yellow, indicating that something important is inside the crematorium, and must be removed.")
+			return
+
+		audible_message("<span class='warning'>You hear a roar as the [src] activates.</span>", 1)
+
+		cremating = TRUE
+		locked = TRUE
+		update_icon()
+
+		for(var/mob/living/M in contents)
+			admin_attack_log(A, M, "Began cremating their victim.", "Has begun being cremated.", "began cremating")
+			if(iscarbon(M))
+				var/mob/living/carbon/C = M
+				for(var/I, I < 60, I++)
+
+					if(C.stat >= UNCONSCIOUS || !(C in contents)) //In case we die or are removed at any point.
+						cremating = 0
+						update_icon()
+						break
+
+					sleep(0.5 SECONDS)
```

Comment:
You know what's coming, right? Have to check we are not qdeleted, are cremating, are not open, etc. At least the qdeleted.